homeassistant/components/modern_forms/fan.py

```diff
@@ -0,0 +1,180 @@
+        self._attr_unique_id = f"{self.coordinator.data.info.mac_address}_fan"
+
+    @property
+    def supported_features(self) -> int:
+        """Flag supported features."""
+        return SUPPORT_DIRECTION | SUPPORT_SET_SPEED
+
+    @property
+    def percentage(self) -> int | None:
+        """Return the current speed percentage."""
+        percentage = 0
+        if bool(self.coordinator.data.state.fan_on):
+            percentage = ranged_value_to_percentage(
+                self.SPEED_RANGE, self.coordinator.data.state.fan_speed
+            )
+        return percentage
+
+    @property
+    def current_direction(self) -> str:
+        """Return the current direction of the fan."""
+        return self.coordinator.data.state.fan_direction
+
+    @property
+    def speed_count(self) -> int:
+        """Return the number of speeds the fan supports."""
+        return int_states_in_range(self.SPEED_RANGE)
+
+    @property
+    def is_on(self) -> bool:
+        """Return the state of the fan."""
+        return bool(self.coordinator.data.state.fan_on)
+
+    @modernforms_exception_handler
+    async def async_set_direction(self, direction: str) -> None:
+        """Set the direction of the fan."""
+        await self.coordinator.modernforms.fan(direction=direction)
+
+    @modernforms_exception_handler
+    async def async_set_percentage(self, percentage: int) -> None:
+        """Set the speed percentage of the fan."""
+        if percentage > 0:
+            await self.async_turn_on(percentage=percentage)
+        else:
+            await self.async_turn_off()
+
+    @modernforms_exception_handler
+    async def async_turn_on(
+        self,
+        speed: int | None = None,
+        percentage: int | None = None,
+        preset_mode: int | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        data = {OPT_ON: FAN_POWER_ON}
+
+        if percentage:
+            data[OPT_SPEED] = round(
+                percentage_to_ranged_value(self.SPEED_RANGE, percentage)
+            )
+        await self.coordinator.modernforms.fan(**data)
+
+    @modernforms_exception_handler
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn the fan off."""
+        await self.coordinator.modernforms.fan(on=FAN_POWER_OFF)
+
+    @modernforms_exception_handler
+    async def async_set_fan_sleep_timer(
+        self,
+        sleep_time: int,
+    ) -> None:
+        """Set a Modern Forms light sleep timer."""
+        await self.coordinator.modernforms.fan(sleep=sleep_time * 60)
+
+    @modernforms_exception_handler
+    async def async_clear_fan_sleep_timer(
+        self,
+    ) -> None:
+        """Clear a Modern Forms fan sleep timer."""
+        await self.coordinator.modernforms.fan(sleep=CLEAR_TIMER)
+
+
+@callback
+def async_update_fan(
+    entry: ConfigEntry,
+    coordinator: ModernFormsDataUpdateCoordinator,
+    current: dict[str, ModernFormsFanEntity],
+    async_add_entities,
```

Comment:
Partially typed, if typing, please type the full signature.

`async_add_entities: AddEntitiesCallback`